src/axelrod_dojo/archetypes/hmm.py

```diff
@@ -0,0 +1,255 @@
+        self.transitions_D = self.mutate_rows(
+            self.transitions_D, self.mutation_probability)
+        self.emission_probabilities = mutate_row(
+            self.emission_probabilities, self.mutation_probability)
+        if random.random() < self.mutation_probability / 10:
+            self.initial_action = self.initial_action.flip()
+        if random.random() < self.mutation_probability / (10 * self.num_states):
+            self.initial_state = randrange(self.num_states)
+        # Change node size?
+
+    @staticmethod
+    def crossover_rows(rows1, rows2):
+        num_states = len(rows1)
+        crosspoint = randrange(num_states)
+        new_rows = copy_lists(rows1[:crosspoint])
+        new_rows += copy_lists(rows2[crosspoint:])
+        return new_rows
+
+    @staticmethod
+    def crossover_weights(w1, w2):
+        crosspoint = random.randrange(len(w1))
+        new_weights = list(w1[:crosspoint]) + list(w2[crosspoint:])
+        return new_weights
+
+    def crossover(self, other):
+        # Assuming that the number of states is the same
+        t_C = self.crossover_rows(self.transitions_C, other.transitions_C)
+        t_D = self.crossover_rows(self.transitions_D, other.transitions_D)
+        emissions = self.crossover_weights(
+            self.emission_probabilities, other.emission_probabilities)
+        return HMMParams(self.num_states, self.mutation_probability,
+                         t_C, t_D, emissions,
+                         self.initial_state, self.initial_action)
+
+    @staticmethod
+    def repr_rows(rows):
+        ss = []
+        for row in rows:
+            ss.append("_".join(list(map(str, row))))
+        return "|".join(ss)
+
+    def __repr__(self):
+        return "{}:{}:{}:{}:{}".format(
+            self.initial_state,
+            self.initial_action,
+            self.repr_rows(self.transitions_C),
+            self.repr_rows(self.transitions_D),
+            self.repr_rows([self.emission_probabilities])
+        )
+
+    @classmethod
+    def parse_repr(cls, s):
+        def parse_vector(line):
+            row = line.split('_')
+            row = list(map(float, row))
+            return row
+        def parse_matrix(sm):
+            rows = []
+            lines = sm.split('|')
+            for line in lines:
+                rows.append(parse_vector(line))
+            return rows
+        lines = s.split(':')
+        initial_state = int(lines[0])
+        initial_action = Action.from_char(lines[1])
+        t_C = parse_matrix(lines[2])
+        t_D = parse_matrix(lines[3])
+        p = parse_vector(lines[4])
+        num_states = len(t_C)
+        return cls(num_states=num_states,
+                   transitions_C=t_C,
+                   transitions_D=t_D,
+                   emission_probabilities=p,
+                   initial_state=initial_state,
+                   initial_action=initial_action)
+
+    def receive_vector(self, vector):
```

Comment:
Can you combine the serialization / deserialization methods? The functionality of the `vector_to_instance` and `receive_vector` methods are covered by the repr class methods on the parameter class. I think the class methods contain all the logic needed to unpack the serialized parameters (so I don't think the quadratic equation is needed).

The classmethods could be improved of course so feel free to do so.

Reply:
Possible I'm missing something, but the parse_repr method parses a string and relies on the special character delimiters.  Here I need to parse a list of floats (what the PSO uses), and without the special characters, I don't see a natural way to delimit this...

Reply:
Ok, I see how the vector functions are being used now with the PSO algorithm. Rather than have `receive_vector` and `vector_to_instance` as instance methods can you make these methods classmethods that return new objects, and name them something like `to_vector` and `from_vector`? Please also explain in the docstring why the additional representation exists (for the PSO algorithm).

Also, I think that `create_vector_bounds` isn't naturally a method of the parameter class, it need not care about an implementation detail of the PSO algorithm. Rather `create_vector_bounds` could be a standalone function (or a staticmethod) that just relies on `N` as a parameter or the size of the input vector to the PSO algorithm.

Finally, consider if it makes sense to supply `N` as a parameter to `vector_to_instance` rather than computing it from the quadratic equation. Are you sure that the chosen root is always the correct one?

Reply:
I will look into all of this.  I'm sure the chosen root is the correct one; the other must be negative.

Reply:
Alright.  I thought about this more, and I agree that what I had before was a sort-of hollow pass-through that didn't really need to live in the class.  On revisiting it and looking at what Gambler is trying to do, I think it makes sense to adjust `receive_vector` to not just store the passed vector in the class as a vector, but to actually deserialize there and store into the appropriate class variables.  When you do this, you can then use the the HMMParams instance like a normally-loaded instance of HMMParams.  You can see this where I changed my test file, instead of running the `read_vector` function to deserialize a `xopt` and put those parameters into an HMMParams class, I can read `xopt` (the vector) right into HMMParams using the `receive_vector` method.  If you do this, then the `vector_to_instance` method is already implemented as the `player` method.  [So I changed the name in PSO and Gambler implementations to `player` to be consistent.]  Finally recognizing `receive_vector` and `create_vector_bounds` _as_ instance methods, you can just use `self.num_states` in these functions instead of using a quadratic.  [This is analogous to using `self.pattern` in the Gambler `create_vector_bounds.]

Then I made all these same changes to FSM.